### PR TITLE
Added InternalsUsedInAttribute. Annotated internal types used in FSharp with the attribute.

### PR DESCRIPTION
**Pull request by TIHan**
This adds a new attribute, `InternalsUsedIn`. Its purpose is to annotate internal types that are used in other assemblies/environments/languages/etc through use of `InternalsVisibleTo`.

The PR also annotates all internal types used in F# with the attribute. This is the first step in ensuring that changes made to these types will not break F# in the future.

Also, here is the current list of internal types used as of today for F#:
```
Microsoft.CodeAnalysis.Classification.ClassificationTags
Microsoft.CodeAnalysis.FindUsages.IFindUsagesContext
Microsoft.CodeAnalysis.Editor.ExportContentTypeLanguageServiceAttribute
Microsoft.CodeAnalysis.Editor.ExportBraceMatcherAttribute
Microsoft.CodeAnalysis.Editor.IBraceMatcher
Microsoft.CodeAnalysis.Editor.INavigationBarItemService
Microsoft.CodeAnalysis.Editor.NavigationBarItem
Microsoft.CodeAnalysis.Editor.IGoToDefinitionService
Microsoft.CodeAnalysis.Editor.BraceMatchingResult
Microsoft.CodeAnalysis.Editor.IContentTypeLanguageService
Microsoft.CodeAnalysis.Editor.IEditorClassificationService
Microsoft.CodeAnalysis.Editor.IEditorFormattingService
Microsoft.CodeAnalysis.Editor.InlineRenameLocation
Microsoft.CodeAnalysis.Editor.InlineRenameReplacement
Microsoft.CodeAnalysis.Editor.IInlineRenameReplacementInfo
Microsoft.CodeAnalysis.Editor.IInlineRenameLocationSet
Microsoft.CodeAnalysis.Editor.IInlineRenameInfo
Microsoft.CodeAnalysis.Editor.IEditorInlineRenameService
Microsoft.CodeAnalysis.Editor.IndentationResult
Microsoft.CodeAnalysis.Editor.ISynchronousIndentationService
Microsoft.CodeAnalysis.Editor.Shared.Utilities.ClassificationTypeMap
Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Presentation.SignatureHelpClassifier
Microsoft.CodeAnalysis.Editor.Implementation.Debugging.BreakpointResolutionResult
Microsoft.CodeAnalysis.Editor.Implementation.Debugging.DebugDataTipInfo
Microsoft.CodeAnalysis.Editor.Implementation.Debugging.DebugLocationInfo
Microsoft.CodeAnalysis.Editor.Implementation.Debugging.IBreakpointResolutionService
Microsoft.CodeAnalysis.Editor.Implementation.Debugging.ILanguageDebugInfoService
Microsoft.CodeAnalysis.Editor.FindUsages.IFindUsagesService
Microsoft.CodeAnalysis.Editor.Shared.Extensions.DependencyObjectExtensions
Microsoft.CodeAnalysis.Glyph
Microsoft.CodeAnalysis.GlyphTags
Microsoft.CodeAnalysis.DocumentSpan
Microsoft.CodeAnalysis.Structure.BlockSpan
Microsoft.CodeAnalysis.Structure.BlockStructure
Microsoft.CodeAnalysis.Structure.BlockStructureService
Microsoft.CodeAnalysis.SignatureHelp.ExportSignatureHelpProviderAttribute
Microsoft.CodeAnalysis.SignatureHelp.ISignatureHelpProvider
Microsoft.CodeAnalysis.SignatureHelp.SignatureHelpItem
Microsoft.CodeAnalysis.SignatureHelp.SignatureHelpItems
Microsoft.CodeAnalysis.SignatureHelp.SignatureHelpParameter
Microsoft.CodeAnalysis.SignatureHelp.SignatureHelpTriggerInfo
Microsoft.CodeAnalysis.SignatureHelp.SignatureHelpTriggerReason
Microsoft.CodeAnalysis.Shared.Options.ServiceFeatureOnOffOptions
Microsoft.CodeAnalysis.Notification.INotificationService
Microsoft.CodeAnalysis.Notification.NotificationSeverity
Microsoft.CodeAnalysis.Navigation.IDocumentNavigationService
Microsoft.CodeAnalysis.Navigation.INavigableItem
Microsoft.CodeAnalysis.Navigation.NavigationOptions
Microsoft.CodeAnalysis.NavigateTo.INavigateToSearchResult
Microsoft.CodeAnalysis.NavigateTo.INavigateToSearchService
Microsoft.CodeAnalysis.NavigateTo.NavigateToMatchKind
Microsoft.CodeAnalysis.FindUsages.DefinitionItem
Microsoft.CodeAnalysis.FindUsages.SourceReferenceItem
Microsoft.CodeAnalysis.DocumentHighlighting.HighlightSpanKind
Microsoft.CodeAnalysis.DocumentHighlighting.HighlightSpan
Microsoft.CodeAnalysis.DocumentHighlighting.DocumentHighlights
Microsoft.CodeAnalysis.DocumentHighlighting.IDocumentHighlightsService
Microsoft.CodeAnalysis.Diagnostics.DocumentDiagnosticAnalyzer
Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerCategory
Microsoft.CodeAnalysis.Diagnostics.DiagnosticCategory
Microsoft.CodeAnalysis.Diagnostics.DiagnosticCustomTags
Microsoft.CodeAnalysis.Diagnostics.IBuiltInAnalyzer
Microsoft.CodeAnalysis.Diagnostics.IDiagnosticAnalyzerService
Microsoft.CodeAnalysis.Completion.CommonCompletionItem
Microsoft.CodeAnalysis.Completion.CommonCompletionProvider
Microsoft.CodeAnalysis.Completion.CommonCompletionUtilities
Microsoft.CodeAnalysis.Completion.CompletionOptions
Microsoft.CodeAnalysis.Completion.FileSystemCompletionHelper
Microsoft.CodeAnalysis.CommentSelection.CommentSelectionInfo
Microsoft.CodeAnalysis.CommentSelection.ICommentSelectionService
Microsoft.CodeAnalysis.PatternMatching.PatternMatcher
Microsoft.CodeAnalysis.PatternMatching.PatternMatch
Microsoft.CodeAnalysis.PatternMatching.PatternMatchKind
Microsoft.VisualStudio.LanguageServices.ProjectSystem.IWorkspaceProjectContext
Microsoft.VisualStudio.LanguageServices.ProjectSystem.IWorkspaceProjectContextFactory
Microsoft.VisualStudio.LanguageServices.Implementation.TaskList.ProjectExternalErrorReporter
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.AbstractProject
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.IVisualStudioHostDocument
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioMetadataReference
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioMetadataReferenceManager
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProjectTracker
Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl
Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractLanguageService
Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractLanguageService`2
Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage`2
Microsoft.VisualStudio.LanguageServices.Implementation.F1Help.IHelpContextService
Microsoft.VisualStudio.LanguageServices.Implementation.DebuggerIntelliSense.AbstractDebuggerIntelliSenseContext

// Auto-gen, can't annotate
Microsoft.CodeAnalysis.Editor.EditorFeaturesResources
```

// cc @tmat

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Was a tool written for this.  We should likely get hte same information for TS and Xaml.

**Comment by tmat** [comment]
I believe @TIHan has a tool. Can we perhaps add it to https://github.com/dotnet/roslyn-tools?

**Comment by tmat** [comment]
Actually, for TS and Xaml it should be implemented as a Roslyn analyzer. F# doesn't have analyzers (yet).

**Comment by CyrusNajmabadi** [comment]
> Actually, for TS and Xaml it should be implemented as a Roslyn analyzer

That woudl make a lot of sense!

**Comment by amcasey** [comment]
I would certainly prefer not to do it by hand. :smile:

**Comment by TIHan** [comment]
I have a little tool to gather all of the internal types. However, I manually placed the attribute on the types in Roslyn. It was about ~90 types in total and only took around 20-25 minutes to do it. I figured that was quicker than trying to build a tool to do it for me.

**Comment by Pilchie** [comment]
Note that Xaml doesn't have IVT's to outside of this repo, except in the MS.VS.LS.Xaml assembly so that they can more closely control what gets exposed externally.

**Comment by mgoertz-msft** [comment]
Yeah, I'm surprised that nobody has started following our example yet. This has worked really well for us. We have never been broken by an internals change in Roslyn.

**Comment by tmat** [comment]
@amcasey @TIHan do you think you could use the same approach as Xaml? That is define some services in a separate assembly built from Roslyn repo and then consume only those services w/o access to other Roslyn internals? I'm little worried it might result in a quite a bit of duplicate code.

**Comment by amcasey** [comment]
@tmat, presumably, that will depend on what our dependencies turn out to be.
@TIHan I agree that the audit is the hard part.  I'm fine creating a manual Roslyn PR decorating the types.  Once we've done that, I assume someone will provide an analyzer that rejects dependencies on un-attributed internals?

**Comment by TIHan** [comment]
@tmat That approach would have been a lot better if we were just starting out. Right now we are so tightly coupled to internals, it would take quite a bit of effort to bring those out as services. It would be a much more robust solution IMO, but it also means we would have to update Roslyn in order to bring functionality to F# tooling (Right now we just add functionality to directly to FSharp.Editor in the F# repo).

The best solution is to design public APIs either in Roslyn or in VS based on what we use. I understand this process is very time consuming and probably not happening anytime soon, but I want to mention it.

**Comment by TIHan** [comment]
@amcasey I think that was the plan; that a analyzer would be created to understand the attributes.

I'm also curious where Xaml, TypeScript and F# intersect in their type usage.

**Comment by tmat** [comment]
It seems to me that the attribute approach is the most viable way forward, given the current state. It also seems more flexible. We can trivially add/remove attributes without writing a lot of delegating code, possibly in multiple copies for APIs used by multiple teams.

Once we annotate all the types we'll also immediately see which APIs do we need to make public. For this to work we need two analyzers:
1) One running in Roslyn that guards changes in the annotated APIs -- this could just be a tweak in the existing public API guard.
2) One running in TS code-base restricting their usage to annotated APIs.

**Comment by amcasey** [comment]
@tmat we might want a third analyzer/tool to detect attributed APIs that are no longer consumed.

**Comment by tmat** [comment]
@amcasey Yeah, that would be useful. Could be part of the second analyzer.

**Comment by tmat** [comment]
@CyrusNajmabadi That might be true for some API but certainly not for all.
In fact, we have started redesigning APIs around AbstractProject significantly. We've also made changes in other APIs during those 5 years that required synchronized update across TS and Roslyn. On multiple occasions we changed API without updating TS which resulted to a break. So I would not say these APIs are that stable.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi That might be true for some API but certainly not for all.

Of course.  I wasn't tryign to imply otherwise :)  This would only be for the small, safe service-oriented APIs.  

>  So I would not say these APIs are that stable.

There are several that are.  And even the APIs that changed i would consider *stable* in the sense that they were working properly and actually could have stayed as is.  i.e. when we broke them it was only becuase we wanted to added new features, but we didn't want to introduce a new API.  For example, adding support for options in NavigateTo.  A client of these APIs should *not* have been broken by this as we should have supported both the original API and a new API that had this functionality.  (of course DefaultInterfaceMethods would have also made this ok).  

In other words: for APIs that may go through major upheaval, i would definitely not do this (i.e. Project System stuff).  For APIs that may *naturally* evolve, but which do so in very basic and well-understood ways (i.e. adding a new parameter for new functionality), then we should do this.  i.e. if we would have used a DIM to bridge original and current functionality, then this is a suitable candidate for making public.

**Comment by zpodlovics** [comment]
@tmat @TIHan @Krzysztof-Cieslak A fresh F# Analyzer development started here https://github.com/Krzysztof-Cieslak/FSharp.Analyzers.SDK and here https://github.com/Krzysztof-Cieslak/FSharp.Analyzers.Sample

**Comment by tmat** [comment]
test this please

**Comment by CyrusNajmabadi** [comment]
I'm going to close this out.  It looks like we don't intend to take this (And we have hte EA pattern for the supported mechanism for F#).  If there is disagreement on this resolution, lmk and we can determine what to do next.
